rbc/nrt.py

```diff
@@ -0,0 +1,670 @@
+
+            dtor = self._get_from_meminfo(mi, "dtor")
+            not_null = cgutils.is_not_null(builder, dtor)
+            with cgutils.if_likely(builder, not_null):
+                dtor_fn = builder.bitcast(dtor, self.NRT_DTOR_FUNCTION.as_pointer())
+                data = self._get_from_meminfo(mi, "data")
+                size = self._get_from_meminfo(mi, "size")
+                dtor_info = self._get_from_meminfo(mi, "dtor_info")
+                builder.call(dtor_fn, [data, size, dtor_info])
+
+            builder.call(self.NRT_MemInfo_destroy, [mi])
+        builder.ret_void()
+
+    def define_NRT_incref(self, builder, args):
+        # with self.nrt_debug_ctx():
+        #     pass
+        builder.ret_void()
+
+    def define_NRT_decref(self, builder, args):
+        # with self.nrt_debug_ctx():
+        #     pass
+        builder.ret_void()
+
+    def define_NRT_MemInfo_alloc_safe_aligned(self, builder, args):
+        # Just call the non-aligned version for now
+        with self.nrt_debug_ctx():
+            [size, align] = args
+            ptr = builder.call(self.NRT_MemInfo_alloc_safe, [size])
+            self.NRT_Debug("ptr=%p\n", ptr)
+        builder.ret(ptr)
+
+    def define_NRT_MemInfo_data_fast(self, builder, args):
+        with self.nrt_debug_ctx():
+            [ptr] = args
+            self.NRT_Debug("ptr=%p\n", ptr)
+            not_null = builder.icmp_signed("!=", ptr, NULL)
+            with builder.if_else(not_null, likely=True) as (then, otherwise):
+                with then:
+                    bb_then = builder.basic_block
+                    mi_ptr = builder.bitcast(
+                        ptr, self.MemInfo_ptr_t, name="meminfo_ptr"
+                    )
+                    data = self._get_from_meminfo(mi_ptr, "data")
+                    self.NRT_Debug("data: %p\n", data)
+                with otherwise:
+                    # there is a specific case where ptr is NULL and doesn't
+                    # crash Numba
+                    bb_else = builder.basic_block
+                    data_null = NULL
+            phi = builder.phi(i8p)
+            phi.add_incoming(data, bb_then)
+            phi.add_incoming(data_null, bb_else)
+        builder.ret(phi)
+
+    def define_NRT_Reallocate(self, builder, args):
+        with self.nrt_debug_ctx():
+            # Because the previously reallocated memory might be freed by
+            # realloc, RBC cannot call it in this function. Therefore, instead
+            # of reallocation, RBC allocates a new block of memory and returns
+            # it to the caller. The caller is responsible for copying memory
+            # from the old buffer into the most recent allocated one."
+            [ptr, size] = args
+            new_ptr = builder.call(self.NRT_Allocate_External, [size, NULL])
+            self.NRT_Debug("ptr=%p size=%zu -> new_ptr=%p", ptr, size, new_ptr)
+        builder.ret(new_ptr)
```

Comment:
Is there any reason why `realloc` can't just be a wrapper around `malloc` + `memcpy` then? That would remove the weirdness about the caller being responsible for copying, which isn't the case for `realloc`. I think it's preferable to keep to the original function's API as much as possible.

Reply:
Ahh that makes sense.